### PR TITLE
Refactor/rewrite all the code

### PR DESCRIPTION
**Pull request by mrkkrp**
The change prepares solid ground for further improvements. This also fixes at least two bugs:

* Proper support for empty lines and comments in build scripts.

* Previously switching to new releases was not atomic despite all the symlink business. This is because it deleted the current link before creating the new one, but it should have created a new link and then use `mv` to replace the old one atomically. The code has been changed to work properly now.

Tests are comping soon, I'll need to figure out what to port and what to add from scratch, so it's still WIP.

### CODE REVIEW

**Comment by mrkkrp** [comment]
@sestrella Perhaps you could take a look and tell me your thoughts so far?

File: src/System/Hapistrano/Core.hs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+-- | Run the given sequence of command. Whether to use SSH or not is
+-- determined from settings contained in the 'Hapistrano' monad
+-- configuration. Commands that return non-zero exit codes will result in
+-- short-cutting of execution.
+
+exec :: forall a. Command a => a -> Hapistrano (Result a)
+exec typedCmd = do
+  Config {..} <- ask
+  let (prog, args) =
+        case configSshOptions of
```
Comment (sestrella) [inline src/System/Hapistrano/Core.hs]: @mrkkrp :+1: nice we could easily run commands locally now.

File: src/System/Hapistrano/Core.hs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+          Just SshOptions {..} ->
+            ("ssh", [sshHost, "-p", show sshPort, cmd])
+      cmd = renderCommand typedCmd
+      hostLabel =
+        case configSshOptions of
+          Nothing              -> "localhost"
+          Just SshOptions {..} -> sshHost ++ ":" ++ show sshPort
+  liftIO $ do
+    printLine hostLabel
+    putStrLn ("$ " ++ cmd)
```
Comment (sestrella) [inline src/System/Hapistrano/Core.hs]: @mrkkrp would be nice to start logging the timestamp as well. We could do that in a different PR.

Comment (sestrella) [inline src/System/Hapistrano/Core.hs]: Looking forward maybe we could use a logging library and configure the log level using the command line. But again that's is just a nice to have. No need to do it on this PR. Actually I'm going to create a GH issue.

File: src/System/Hapistrano/Commands.hs [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+  renderCommand Whoami = "whoami"
+  parseResult Proxy = trim
+
+-- | Specify directory in which to perform another command.
+
+data Cd cmd = Cd (Path Abs Dir) cmd
+
+instance Command cmd => Command (Cd cmd) where
+  type Result (Cd cmd) = Result cmd
+  renderCommand (Cd path cmd) = "cd " ++ quote (fromAbsDir path) ++
```
Comment (mrkkrp) [inline src/System/Hapistrano/Commands.hs]: This is a good idea, gonig to add this.
